feat: write a detailed README for learning a policy in Isaac Lab (#10)
aica-technology/isaac-lab: aica_docs/set_up_reinforcement_learning.md

# Reinforcement Learning with Isaac Lab

This README provides a step-by-step guide for training a Neural Network-based Reinforcement Learning (RL) policy in a simulated environment using Isaac Lab and exporting the trained policy in ONNX format.

Isaac Lab is a modular framework designed to simplify robotics workflows, including reinforcement learning, learning from demonstrations, and motion planning. Built on NVIDIA Isaac Sim, it leverages PhysX simulation to deliver photo-realistic environments and high-performance capabilities. With end-to-end GPU acceleration, Isaac Lab enables faster and more efficient training of RL policies.

# Prerequisites

Before training a new policy, begin by cloning the AICA fork of [Isaac Lab](https://github.com/aica-technology/isaac-lab). 

> Next, refer to the [Isaac Lab Installation](https://isaac-sim.github.io/IsaacLab/main/source/setup/installation/pip_installation.html) guide for comprehensive instructions on installation and developer setup.

Author: Indeed, I added more of high level explanation and a verification step for the users before continuing.

Through iterative interactions, the actor refines its understanding of the environment and learns to achieve the desired task efficiently, guided by the rewards it accumulates. 

Isaac Lab enables **actors** to perform actions and execute policies within simulated environments, supporting thousands of parallel instances. This parallelization significantly accelerates training cycles, making it highly efficient for reinforcement learning tasks.

To set up a Reinforcement Learning environment in Isaac Lab, familiarize yourself with the following topics.

## Asset Management

Assets are objects defined within a 3D scene and can belong to one of the following categories: (1) Articulated Objects, (2) Rigid Objects, or (3) Deformable Objects. These assets are represented in the USD (Universal Scene Description) format. 

Predefined assets configurations are located in the following directory in Isaac Lab Repo:  
`source/extensions/omni.isaac.lab_assets/omni/isaac/lab_assets`.  

> This directory contains a range of manipulator robots, including the Franka Panda, Universal Robot UR5E and UR10, Kinova, uFactory, and Kuka. To define a new asset, an asset configuration file must be created within the predefined directory. This file should reference a corresponding USD file. For detailed instructions on importing a new robot not included in the predefined directory, refer to [Importing a New Asset](https://isaac-sim.github.io/IsaacLab/main/source/how-to/import_new_asset.html).

Author: agreed, I will add the explicit name 👍

### Example of Asset Configuration
Here is an example of defining an articulation configuration to set up an asset in a Reinforcement Learning environment:

```python

> UR10_CFG = ArticulationCfg(

Author: Yep, my reasoning here is that I take snippets of code and explain it rather than a full script as the full script are in the repo. Additionally, I think it's better if I link to the original file for completeness.

> UR10_CFG = ArticulationCfg(

Reviewer: You're not linking to the file though, you could to that on "Here" in the above sentence

- **`actuators`**: Sets up the robot's actuator model, specifying parameters such as velocity limits, effort limits, stiffness, and damping.

In this example, the actuator model is defined using `ImplicitActuatorCfg`. However, actuator models in Isaac Lab can be either implicit or explicit. For more information on configuring actuators, refer to the [Actuators in Isaac Lab](https://isaac-sim.github.io/IsaacLab/main/source/overview/core-concepts/actuators.html).

### Isaac Nucleus
Isaac Nucleus is part of NVIDIA’s Omniverse platform and serves as a central repository for assets and utilities used in Isaac Lab. It provides a comprehensive collection of prebuilt USD assets such as objects, tables, and manipulator robots that greatly simplify scene construction. Additionally, Isaac Nucleus simplifies collaboration by storing and sharing all necessary files in one location, making it easier for multiple users to work together on robotics simulations and environments.

As shown in the example articulation configuration, the relevant USD file is stored in Isaac Nucleus and can be accessed by importing `ISAACLAB_NUCLEUS_DIR` from `omni.isaac.lab.utils.assets` in Isaac Lab.

> Beyond the default assets, AICA has curated a list of additional resources not included in the default Isaac Sim folder on Isaac Nucleus, such as the uFactory xArm 6 and KUKA KR210 robots, which can be made available upon request.

Reviewer: Ok makes sense but seems somewhat weird in this document. I feel we should
- either also make the asset configuration available on request
- or find a way to make clear that even though you said a few lines above that these configurations are there, its not usable without USD file

2. **SKRL**: An open-source modular RL library written in Python, built on top of PyTorch and JAX. SKRL emphasizes modularity, readability, and simplicity, supporting various environments, including NVIDIA Isaac Gym and Omniverse Isaac Gym. It enables simultaneous training of multiple agents with customizable scopes. More details can be found on [SKRL Github](https://github.com/Toni-SM/skrl).

3. **RL Games**: An open-source high-performance reinforcement learning framework implemented in Python, designed for training policies in simulated environments. More details can be found on [RL Games Github](https://github.com/Toni-SM/skrl).

For detailed instructions on running the various RL libraries, refer to [Reinforcement Learning Wrappers](https://isaac-sim.github.io/IsaacLab/main/source/overview/reinforcement-learning/rl_existing_scripts.html).

To prepare a reinforcement learning environment for use with one of the mentioned wrappers, additional configuration files need to be specified. For an example, check out the directory in the Isaac Lab fork:

`source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/manager_based/manipulation/reach`

### Step 1: Create a New Folder  

> Begin by creating a new folder under the predefined directory. The folder name should typically correspond to the RL actor (in this case, the robot being used).  

Author: aha interesting, the actor is the Neural Network that acts upon the Robot, I will re-phrase 👍

    kwargs={
        "env_cfg_entry_point": joint_pos_env_cfg.UR5EJointPosReachEnvCfg,  # Configuration for the environment, inheriting from ManagerBasedRLEnv
        "rl_games_cfg_entry_point": f"{agents.__name__}:rl_games_ppo_cfg.yaml",  # Path to the RL Games configuration file
        "rsl_rl_cfg_entry_point": f"{agents.__name__}.rsl_rl_ppo_cfg:UR5eReachPPORunnerCfg",  # Path to the RSL-RL configuration file
        "skrl_cfg_entry_point": f"{agents.__name__}:skrl_ppo_cfg.yaml",  # Path to the SK-RL configuration file
    }
)
```

This registration code should be included in the **`__init__.py`** file located in the directory specified above.

## Training and Exporting the Reinforcement Learning Policies

After defining the assets, environments, and robot control, creating the necessary training configuration files, and registering the new Gym environment, train a new policy by running the appropriate training script for the Reinforcement Learning Wrappers. Detailed instructions are available [here](https://isaac-sim.github.io/IsaacLab/main/source/overview/reinforcement-learning/rl_existing_scripts.html).

> Once training is complete and the results meet your expectations, export the model in ONNX format. In RSL-RL, this can be easily achieved using **export_policy_as_onnx**. 

Reviewer: Do we need an example here, or where can one find one?

Author: We do, added an example!

# Next Steps to Execute the Policy on the Robot 

> With the ONNX policy in hand, AICA provides an SDK that enables you to take your simulation-trained policy and deploy it onto real hardware. This [README](https://github.com/aica-technology/dynamic-components/tree/main/source/advanced_components/rl_policy_components) offers a step-by-step guide for deploying a learned policy from Isaac Lab Omniverse to various robotic brands and integrating it into any complex application you develop in AICA Studio.

Reviewer: "at hand" or "in hand"? I'm not sure

Author: "at hand" feels more natural I think, I'll change that

> With the ONNX policy in hand, AICA provides an SDK that enables you to take your simulation-trained policy and deploy it onto real hardware. This [README](https://github.com/aica-technology/dynamic-components/tree/main/source/advanced_components/rl_policy_components) offers a step-by-step guide for deploying a learned policy from Isaac Lab Omniverse to various robotic brands and integrating it into any complex application you develop in AICA Studio.

Reviewer: ```suggestion
With the ONNX policy in hand, AICA provides an SDK that enables you to take your simulation-trained policy and deploy it on real hardware. This [README](https://github.com/aica-technology/dynamic-components/tree/main/source/advanced_components/rl_policy_components) offers a step-by-step guide for deploying a learned policy from Isaac Lab Omniverse to various robotic brands and integrating it into any complex application you develop in AICA Studio.
```

> With the ONNX policy in hand, AICA provides an SDK that enables you to take your simulation-trained policy and deploy it onto real hardware. This [README](https://github.com/aica-technology/dynamic-components/tree/main/source/advanced_components/rl_policy_components) offers a step-by-step guide for deploying a learned policy from Isaac Lab Omniverse to various robotic brands and integrating it into any complex application you develop in AICA Studio.

Reviewer: The whole document is written in 3rd person, only the last paragraphs use 2nd person ("your expectations", "you develop"). We might be able to keep it 3rd person no?

Author: Indeed!

# Reinforcement Learning with Isaac Lab

This document provides a step-by-step guide for training a Reinforcement Learning (RL) policy based on Neural Networks (NN) in a simulated environment using Isaac Lab and exporting the trained policy in ONNX format.

Isaac Lab is a modular framework designed to simplify robotics workflows, including Reinforcement Learning, learning from demonstrations, and motion planning. Built on NVIDIA Isaac Sim, it leverages PhysX simulation to deliver photo-realistic environments and high-performance capabilities. With end-to-end GPU acceleration, Isaac Lab enables faster and more efficient training of RL policies.

# Prerequisites

Before training a new policy, begin by cloning the AICA fork of [Isaac Lab](https://github.com/aica-technology/isaac-lab).

> Next, refer to the [Isaac Lab Installation](https://isaac-sim.github.io/IsaacLab/main/source/setup/installation/pip_installation.html) guide for detailed steps on installation and developer setup. Isaac Lab is built on top of Isaac Sim, so first follow the Isaac Sim installation procedure (which can be done through pip). Afterward, install Isaac Lab by adding the necessary development dependencies and running the installation script in the cloned repository (**./isaaclab.sh --install**).

Reviewer: ```suggestion
Next, refer to the [Isaac Lab Installation](https://isaac-sim.github.io/IsaacLab/main/source/setup/installation/pip_installation.html) guide for detailed steps on installation and developer setup. Isaac Lab is built on top of Isaac Sim, so first follow the Isaac Sim installation procedure (which can be done through pip). Afterwards, install Isaac Lab by adding the necessary development dependencies and running the installation script in the cloned repository (`./isaaclab.sh --install`).
```

Through iterative interactions, the actor refines its understanding of the environment and learns to achieve the desired task efficiently, guided by the rewards it accumulates. 

Isaac Lab enables **actors** to perform actions and learn policies within simulated environments, supporting thousands of parallel instances. This parallelization significantly accelerates training cycles, making it highly efficient for RL tasks.

To set up a RL environment in Isaac Lab, familiarize yourself with the following topics.

## Asset Management

Assets are objects defined within a 3D scene and can belong to one of the following categories: (1) Articulated Objects, (2) Rigid Objects, or (3) Deformable Objects. These assets are represented in the USD (Universal Scene Description) format. 

Predefined assets configurations are located in the following directory in the Isaac Lab repository:  
`source/extensions/omni.isaac.lab_assets/omni/isaac/lab_assets`.  

> This directory contains a range of manipulator robots, including the Franka Emika Panda, Universal Robot UR5e and UR10, Kinova JACO2 (7-Dof), JACO2 (6-Dof) and Gen3 (7-Dof), uFactory xArm 6, and Kuka KR210. To define a new asset, an asset configuration file must be created within [lab_assets directory](source/extensions/omni.isaac.lab_assets/omni/isaac/lab_assets). This file should reference a corresponding USD file. For detailed instructions on importing a new robot not included in [lab_assets directory](source/extensions/omni.isaac.lab_assets/omni/isaac/lab_assets), refer to [Importing a New Asset](https://isaac-sim.github.io/IsaacLab/main/source/how-to/import_new_asset.html).

Reviewer: ```suggestion
This directory contains a range of manipulator robots, including the Franka Emika Panda, Universal Robot UR5e and UR10, Kinova JACO2, JACO2 and Gen3, uFactory xArm 6, and Kuka KR210. To define a new asset, an asset configuration file must be created within the [lab_assets directory](source/extensions/omni.isaac.lab_assets/omni/isaac/lab_assets). This file should reference a corresponding USD file. For detailed instructions on importing a new robot not included in the [lab_assets directory](source/extensions/omni.isaac.lab_assets/omni/isaac/lab_assets), refer to [Importing a New Asset](https://isaac-sim.github.io/IsaacLab/main/source/how-to/import_new_asset.html).
```

### Isaac Nucleus
Isaac Nucleus is part of NVIDIA’s Omniverse platform and serves as a central repository for assets and utilities used in Isaac Lab. It provides a comprehensive collection of prebuilt USD assets such as objects, tables, and manipulator robots that greatly simplify scene construction. Additionally, Isaac Nucleus simplifies collaboration by storing and sharing all necessary files in one location, making it easier for multiple users to work together on robotics simulations and environments.

As shown in the example articulation configuration, the relevant USD file is stored in Isaac Nucleus and can be accessed by importing `ISAACLAB_NUCLEUS_DIR` from `omni.isaac.lab.utils.assets` in Isaac Lab.

Beyond the default assets, AICA has curated a list of additional resources not included in the default Isaac Sim folder on Isaac Nucleus, such as the uFactory xArm 6 and KUKA KR210 robots, which can be made available upon request.

## Simulation Environments

> Simulation environments are virtual environments where a RL agent takes actions, observes states, and receives rewards. In Isaac Lab, there are two types of environments:

Reviewer 2: ```suggestion
Simulation environments are virtual environments with which a RL actor interacts to learn a policy. In Isaac Lab, there are two types of environments:
```

1. [Manager-Based RL Environment](https://isaac-sim.github.io/IsaacLab/main/source/tutorials/03_envs/create_manager_rl_env.html)  
2. [Direct RL Environment](https://isaac-sim.github.io/IsaacLab/main/source/tutorials/03_envs/create_direct_rl_env.html)

> Manager-Based RL Environments simplify simulation complexity by using managers that handle tasks such as fetching, updating, and setting the data required by the RL agent (for example, constructing the actor state from observations). These managers are **InteractiveScene**, **ActionManager**, **ObservationManager**, **RewardManager**, **CurriculumManager**, and **EventManager**.  In contrast, Direct-Based RL Environments offer greater flexibility by requiring users to define observations, actions, and rewards directly within the task script.

Reviewer: ```suggestion
Manager-Based RL Environments simplify simulation complexity by using managers that handle tasks such as fetching, updating, and setting the data required by the RL agent (for example, constructing the actor state from observations). These managers are `InteractiveScene`, **ActionManager**, **ObservationManager**, **RewardManager**, **CurriculumManager**, and **EventManager**.  In contrast, Direct-Based RL Environments offer greater flexibility by requiring users to define observations, actions, and rewards directly within the task script.
```
If those are all actual classes, then please with `code block`

In what comes next, Manager-Based RL Environments will be explored in more details. 

### Manager-Based RL Environment

To create your own Manager-Based RL Environment, follow this [tutorial](https://isaac-sim.github.io/IsaacLab/main/source/tutorials/03_envs/create_manager_rl_env.html). Below is a summary of a basic environment configuration class:

```python
class ReachEnvCfg(ManagerBasedRLEnvCfg):
    """Configuration for the end-effector pose tracking environment."""

    # Scene settings
    scene: ReachSceneCfg = ReachSceneCfg(num_envs=4096, env_spacing=2.5)

    # Basic settings
    observations: ObservationsCfg = ObservationsCfg()
    actions: ActionsCfg = ActionsCfg()
    commands: CommandsCfg = CommandsCfg()

    # MDP settings
    rewards: RewardsCfg = RewardsCfg()
    terminations: TerminationsCfg = TerminationsCfg()
    events: EventCfg = EventCfg()
    curriculum: CurriculumCfg = CurriculumCfg()
```

> As shown in the configuration above, **ReachEnvCfg** inherits from **ManagerBasedRLEnvCfg** and sets up the various managers that define the RL environment. The **ReachSceneCfg** specify the assets in the environment such as the ground, robot, and lights as demonstrated in the snippet below:

Reviewer 2: ```suggestion
As shown in the configuration above, **ReachEnvCfg** inherits from **ManagerBasedRLEnvCfg** and sets up the various managers that define the environment. The **ReachSceneCfg** specify the assets in the environment such as the ground, robot, and lights as demonstrated in the snippet below:
```

> As shown in the configuration above, **ReachEnvCfg** inherits from **ManagerBasedRLEnvCfg** and sets up the various managers that define the RL environment. The **ReachSceneCfg** specify the assets in the environment such as the ground, robot, and lights as demonstrated in the snippet below:

Reviewer: Inline code blocks instead of bold (across the whole file)

```python
@configclass
class ReachSceneCfg(InteractiveSceneCfg):
    """Configuration for the scene with a robotic arm."""

    # world
    ground = AssetBaseCfg(
        prim_path="/World/ground",
        spawn=sim_utils.GroundPlaneCfg(),
        init_state=AssetBaseCfg.InitialStateCfg(pos=(0.0, 0.0, 0.0)),
    )

    # robots
    robot: ArticulationCfg = MISSING

    # lights
    light = AssetBaseCfg(
        prim_path="/World/light",
        spawn=sim_utils.DomeLightCfg(color=(0.75, 0.75, 0.75), intensity=2500.0),
    )
```

> The **ObservationsCfg** class specifies the observations passed to the RL agent in the form of a state vector, which can include information such as joint positions, joint velocities, and end-effector pose for robotic arms.

Reviewer 2: ```suggestion
The **ObservationsCfg** class specifies the observations passed to the RL actor in the form of a state vector, which can include information such as joint positions, joint velocities, and end-effector pose for robotic arms.
```

We need to commit to either actor or agent throughout the text

```python
@configclass
class ObservationsCfg:
    """Observation specifications for the MDP."""

    @configclass
    class PolicyCfg(ObsGroup):
        """Observations for the policy group."""

        # Observation terms (order preserved)
        joint_pos = ObsTerm(func=mdp.joint_pos_rel, noise=Unoise(n_min=-0.01, n_max=0.01))
        ee_position = ObsTerm(func=mdp.ee_position_in_robot_root_frame)
        ee_orientation = ObsTerm(func=mdp.ee_rotation_in_robot_root_frame)
        pose_command = ObsTerm(func=mdp.generated_commands, params={"command_name": "ee_pose"})
        
        joint_vel = ObsTerm(func=mdp.joint_vel_rel, noise=Unoise(n_min=-0.01, n_max=0.01))
        actions = ObsTerm(func=mdp.last_action)

        def __post_init__(self):
            self.enable_corruption = True
            self.concatenate_terms = True

    # Observation groups
    policy: PolicyCfg = PolicyCfg()
```

> Each observation term specifies how a particular piece of data, such as joint positions or end-effector orientation, is retrieved, optionally corrupted with noise (for dynamic randomization), and appended to the RL agent’s input vector. The `func` parameter identifies a callback function that supplies these values, while the order of terms in the configuration dictates their arrangement in the final state input.

Reviewer 2: ```suggestion
Each observation term specifies how a particular piece of data, such as joint positions or end-effector orientation, is 1) retrieved, optionally corrupted with noise (for dynamic randomization), and 2) appended to the RL agent’s input vector. The `func` parameter identifies a callback function that supplies these values, while the order of terms in the configuration dictates their arrangement in the final state input.
```

        asset_name="robot", 
        joint_names=[".*"], 
        scale=1.0, 
        use_default_offset=True
    )
```

### Example: Joint Velocity Control
Here, the policy outputs joint velocity setpoints for controlling the robot's movement:

```python
@configclass
class ActionsCfg:
    """Action specifications for the MDP."""

    arm_action: ActionTerm = mdp.JointVelocityActionCfg(
        asset_name="robot", 
        joint_names=[".*"], 
        scale=1.0, 
        use_default_offset=True
    )
```

### Example: Cartesian Twist Control
In this setup, the policy outputs Cartesian twist commands (i.e. velocities in Cartesian space) that are translated into joint commands using an inverse kinematics solver:

```python
@configclass
class ActionsCfg:
    """Action specifications for the MDP."""

    arm_action: ActionTerm = DifferentialInverseKinematicsActionCfg(
        asset_name="robot",
        joint_names=[".*"],
        body_name="tool0",  # End-effector frame name
        controller=DifferentialIKControllerCfg(
            command_type="pose", 
            use_relative_mode=True, 
            ik_method="dls"
        ),
        scale=1.0,
    )
```

> Below is a revised, more fluent version:
> 
> ---

Reviewer 2: ```suggestion
```

    kwargs={
        "env_cfg_entry_point": joint_pos_env_cfg.UR5EJointPosReachEnvCfg,  # Configuration for the environment, inheriting from ManagerBasedRLEnv
        "rl_games_cfg_entry_point": f"{agents.__name__}:rl_games_ppo_cfg.yaml",  # Path to the RL Games configuration file
        "rsl_rl_cfg_entry_point": f"{agents.__name__}.rsl_rl_ppo_cfg:UR5eReachPPORunnerCfg",  # Path to the RSL-RL configuration file
        "skrl_cfg_entry_point": f"{agents.__name__}:skrl_ppo_cfg.yaml",  # Path to the SK-RL configuration file
    }
)
```

This registration code should be included in the **`__init__.py`** file located in the directory specified above.

## Training and Exporting the Reinforcement Learning Policies

After defining the assets, environments, and robot control, creating the necessary training configuration files, and registering the new Gym environment, train a new policy by running the appropriate training script for the Reinforcement Learning Wrappers. Detailed instructions are available [here](https://isaac-sim.github.io/IsaacLab/main/source/overview/reinforcement-learning/rl_existing_scripts.html).

> Below is a cleaned-up, more fluent version:
> 
> ---

Reviewer 2: ```suggestion
```

        """Observations for the policy group."""

        # Observation terms (order preserved)
        joint_pos = ObsTerm(func=mdp.joint_pos_rel, noise=Unoise(n_min=-0.01, n_max=0.01))
        ee_position = ObsTerm(func=mdp.ee_position_in_robot_root_frame)
        ee_orientation = ObsTerm(func=mdp.ee_rotation_in_robot_root_frame)
        pose_command = ObsTerm(func=mdp.generated_commands, params={"command_name": "ee_pose"})

        joint_vel = ObsTerm(func=mdp.joint_vel_rel, noise=Unoise(n_min=-0.01, n_max=0.01))
        actions = ObsTerm(func=mdp.last_action)

        def __post_init__(self):
            self.enable_corruption = True
            self.concatenate_terms = True

    # Observation groups
    policy: PolicyCfg = PolicyCfg()
```

Each observation term specifies how a particular piece of data, such as joint positions or end-effector orientation,
is 1) retrieved, optionally corrupted with noise (for dynamic randomization), and 2) appended to the RL agent’s input
vector. The `func` parameter identifies a callback function that supplies these values, while the order of terms in the
configuration dictates their arrangement in the final state input.

> The `ActionsCfg` class defines how the outputs of the Neural Network are translated into actions applied in the

Reviewer 2: ```suggestion
The `ActionsCfg` class defines how the outputs of the neural network are translated into actions applied in the
```

I don't see a reason to capitalize `Neural Network` (or `Reinforcement Learning` for that matter) throughout the text, except for when it's in titles which ok given the convention you are using.

   A Robotic arm learning to reach a specified end-effector position and orientation.

2. [Lift Environment](../source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/manager_based/manipulation/lift/lift_env_cfg.py)  
   A Robotic arm learning to lift an object within the workspace.

3. [Stack Environment](../source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/manager_based/manipulation/stack/stack_env_cfg.py)  
   A Robotic arm learning to stack multiple objects.

For more details on complex reward terms, such as triggering penalties when an object falls out of the workspace, and on
curriculum learning refer to the official
[Isaac Lab Manager-Based RL Environment](https://isaac-sim.github.io/IsaacLab/main/source/tutorials/03_envs/create_manager_rl_env.html)
documentation.

## Reinforcement Learning Workflows

Several open-source RL frameworks facilitate policy learning and optimization. Isaac Lab offers wrappers for various RL
libraries, translating environment data into the appropriate formats for each library's functions. The three supported
libraries are:

> 1. `RSL-RL`: Developed by the Robotic Systems Lab at ETH Zürich, RSL-RL is a fast and straightforward implementation of

Reviewer 2: ```suggestion
1. `RSL-RL`: Developed by the Robotic Systems Lab (RSL) at ETH Zürich, RSL-RL is a fast and straightforward implementation of
```

   and simplicity, supporting various environments, including NVIDIA Isaac Gym and Omniverse Isaac Gym. It enables
   simultaneous training of multiple agents with customizable scopes. More details can be found on
   [SKRL Github](https://github.com/Toni-SM/skrl).

3. `RL Games`: An open-source high-performance RL framework designed for training policies in simulated environments.
   More details can be found on [RL Games Github](https://github.com/Toni-SM/skrl).

For detailed instructions on running the various RL libraries, refer to
[Reinforcement Learning Wrappers](https://isaac-sim.github.io/IsaacLab/main/source/overview/reinforcement-learning/rl_existing_scripts.html).

To prepare a RL environment for use with one of the mentioned wrappers, additional configuration files need to be
specified. For an example, check out the
[reach directory](../source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/manager_based/manipulation/reach) in
the Isaac Lab repository.

### Step 1: Create a New Folder

Begin by creating a new folder. Its location depends on the simulation environment and the specific task being

> performed. For example, for a Manager-Based RL robotic arm manipulation reach task, an appropriate location for that

Reviewer 2: ```suggestion
performed. For example, for a Manager-Based RL robotic arm manipulation reach task, an appropriate location for that
```

Here, for example, I'm ok(ish) with capitalizing `Manager-Based RL` only because NVidia is consistently going by this convention.

        "skrl_cfg_entry_point": f"{agents.__name__}:skrl_ppo_cfg.yaml",  # Path to the SK-RL configuration file
    }
)
```

This registration code should be included in the `__init__.py` file located in the directory specified above.

## Training and Exporting the Reinforcement Learning Policies

After defining the assets, environments, and robot control, creating the necessary training configuration files, and
registering the new Gym environment, train a new policy by running the appropriate training script for the Reinforcement
Learning Wrappers. Detailed instructions are available
[here](https://isaac-sim.github.io/IsaacLab/main/source/overview/reinforcement-learning/rl_existing_scripts.html).

Once training is complete, you can export your model to ONNX format. In RSL-RL, this is straightforward using
`export_policy_as_onnx`. For other libraries or custom policies, you can use the PyTorch ONNX exporter as described in
the [official documentation](https://pytorch.org/docs/stable/onnx.html). Note that extra attention is required for
recurrent actor-critic models Below are example snippets showing how to export both recurrent and feed-forward
actor-critic models:

> `Recurrent model (Memory Based Models)`:

Reviewer: ```suggestion
### Recurrent model (Memory Based Models):
```

```python
obs = torch.zeros(1, self.rnn.input_size)
h_in = torch.zeros(self.rnn.num_layers, 1, self.rnn.hidden_size)
c_in = torch.zeros(self.rnn.num_layers, 1, self.rnn.hidden_size)

actions, h_out, c_out = self(obs, h_in, c_in)

torch.onnx.export(
    self,
    (obs, h_in, c_in),
    os.path.join(path, filename),
    export_params=True,
    opset_version=11,
    verbose=self.verbose,
    input_names=["obs", "h_in", "c_in"],
    output_names=["actions", "h_out", "c_out"],
    dynamic_axes={},
)
```

> `Feed-forward model (Common Case)`:

Reviewer: ```suggestion
### Feed-forward model (Common Case):
```

# Reinforcement Learning with Isaac Lab

> This document provides a step-by-step guide for training a reinforcement learning (RL) policy based on neural networks

Reviewer 2: ```suggestion
This document provides a step-by-step guide for training a Reinforcement Learning (RL) policy based on neural networks
```

first time the abbreviation is defined it needs to be capitalized
